sematic/plugins/building/docker_builder.py

```diff
@@ -339,30 +388,170 @@ def _find_build_config_files(script_path: str) -> List[str]:
+) -> Tuple[Image, ImageURI]:
     """
     Builds the container image to use, according to the build configuration, and returns
     an `ImageURI` that identifies it.
 
+    Parameters
+    ----------
+    target: str
+        The path to the pipeline target to launch; the built image must support this
+        target's execution.
+    build_config: BuildConfig
+        The configuration that controls the image build.
+    docker_client: docker.DockerClient
+        The client to use for executing the operations.
+
+    Returns
+    -------
+    Tuple[Image, ImageURI]:
+        The build container image and a URI that identifies the image.
+
     Raises
     ------
     BuildError:
-        There was an error when executing the specified build script.
+        There was an error when building the image.
     BuildConfigurationError:
         There was an error when validating the specified build configuration.
+    SystemExit:
+        A subprocess exited with an unexpected code.
     """
-    if build_config.base_uri is not None:
-        # TODO: honor configurations in build_config.build
-        return build_config.base_uri
+    effective_base_uri = build_config.base_uri
 
-    if build_config.image_script is None:
-        # appease mypy
-        raise BuildConfigurationError(
-            "Exactly one of `image_script` and `base_uri` must be specified!"
+    if build_config.image_script is not None:
+        effective_base_uri = _execute_build_script(
+            target=target, image_script=build_config.image_script
+        )
+
+    # appease mypy
+    assert effective_base_uri is not None
+
+    return _build_image_from_base(
+        target=target,
+        effective_base_uri=effective_base_uri,
+        build_config=build_config,
+        docker_client=docker_client,
+    )
+
+
+def _build_image_from_base(
+    target: str,
+    effective_base_uri: ImageURI,
+    build_config: BuildConfig,
+    docker_client: docker.DockerClient,  # type: ignore
+) -> Tuple[Image, ImageURI]:
+    """
+    Builds the container image to use by adding layers to an existing base image.
+    """
+    logger.info("Building image starting from base: %s", effective_base_uri)
+
+    dockerfile_contents = _generate_dockerfile_contents(
+        base_uri=effective_base_uri,
+        source_build_config=build_config.build,
+        target=target if build_config.base_uri is not None else None,
+    )
+    logger.info("Using Dockerfile:\n%s", dockerfile_contents)
+
+    # we have to create a tmp dockerfile and pass it instead of using the `fileobj` option
+    # of the docker_client, because it does not work with contexts, so operations like
+    # COPY do not work, as there exists no working dir context to copy them from
+    # API: https://docker-py.readthedocs.io/en/stable/api.html#module-docker.api.build
+    # Reported unresolved issue: https://github.com/docker/docker-py/issues/2105
+    with tempfile.NamedTemporaryFile(mode="wt", delete=True) as dockerfile:
```

Comment:
I believe that if you write the dockerfile to a new file every single time, then Docker will not be able to cache steps--even expensive ones like pip installing the requirements. Here's an approach that should enable layer caching to work as expected:

1. hash the dockerfile contents
2. get the tempdir for the system, and construct the filename as `{tempdir}/Dockerfile.<content hash>`
3. If the filename already exists, don't overwrite it. If it doesn't exist, write the contents to it.
4. Do the build using the dockerfile with the constructed filename.

Since the ephemeral dockerfiles were written to the system's tempdir, they will persist across process executions and be reusable for the docker daemon's caching. But they are still not creating junk that will bother the user (temp dir can be cleaned as usual to clear up space).

Reply:
Great! I actually didn't check; this comment came from experience creating a CLI tool at a previous job that also generated a dockerfile to build user's code. Perhaps new versions of docker handle caches even if the dockerfile timestamp is newer! Regardless, if caching is working then all is well!